adr/2024-12-19-offer-html-alternative-to-our-pdf-standard-document.md

```diff
@@ -0,0 +1,153 @@
+---
+title: [A11y] Offer HTML alternative to our pdf standard documents
+date: 2024-12-19
+area: accessibility
+tags: [a11y, pdf, html]
+---
+
+## Context
+To ensure compliance with Web Content Accessibility Guidelines (WCAG), we aim to make Shopware 6's standard document solution A11y-compliant. This will enable continued use of the solution while adhering to legal requirements and making documents accessible to customers with disabilities.
+
+Currently, our PDF generation library, DomPDF, does not meet accessibility standards, which presents a significant challenge for PDF accessibility.
+
+## Decision
+In addition to the existing PDF documents generated with DomPDF, we will provide standard document types in an accessible HTML format that fulfills A11y requirements.
+
+### Affected areas
+We will integrate HTML document support in the following areas:
+1. **Order Detail Page (Admin)**:
+    - Preview options:
+        - PDF preview
+        - HTML preview
+
+2. **Order Detail Page (Storefront)**:
+    - Download options:
+        - PDF download
+        - HTML download
+
+3. **Document Creation**:
+    - **Flow Builder**: Automatically generates both PDF and HTML documents.
+    - **Bulk Edit**: Automatically generates both PDF and HTML documents.
+
+4. **Email Attachments**:
+    - Include the new HTML document format.
+
+5. **E-Invoicing**:
+    - Enable document export as HTML.
+
+### Technical implementation
+* To support this, we will modify the document entity to store media IDs for both PDF and HTML documents:
+
+```php
+$this->addColumn(
+    connection: $connection,
+    table: 'document',
+    column: 'document_media_file_ids',
+    type: 'JSON'
+);
+```
+
+* In class `Shopware\Core\Checkout\Document\Controller\DocumentController` we will update method `downloadDocument` (`'/api/_action/document/{documentId}/{deepLinkCode}'`) to get a new query parameter `fileType` which can be `pdf` or `html`. If the type is `html` we will return the HTML document.
```

Comment:
The route and controller are not very interesting. May we add a section on how we protect HTML files and store all those files?